components/change-component-factory.js

```diff
@@ -0,0 +1,130 @@
+ * indexed payload and non-indexed payload. However, when payload is received on
+ * a non-indexed port, change event is only triggered with other non-indexed payloads.
+ *
+ * Usage:
+ *   componentFactory({
+ *     inPorts: {input:{indexBy:'vnid'}},
+ *     outPorts: {output:{indexBy:'vnid'}},
+ *     onchange: function({input:payload,output:payload})
+ *   });
+ */
+module.exports = function(def){
+    return componentFactory(_.defaults({
+        outPorts: _.mapObject(def.outPorts, indexData),
+        inPorts: _.mapObject(def.inPorts, indexDataAndFireChange)
+    }, def));
+};
+
+/**
+ * Adds an ondata event handler (if one does not already exist) to the port
+ * definition that will index the payload using an indexBy property on the port.
+ * @param port port definition
+ * @param portName port name
+ */
+function indexData(port, portName) {
+    return _.extend({
+        ondata: _.partial(indexPayload, createIndexBy(port), portName)
+    }, port);
+}
+
+/**
+ * Adds an ondata event handler (if one does not already exist) to the port
+ * definition that will index the payload using an indexBy property on the port
+ * and conditionally trigger a change event with the matching payload.
+ * @param port port definition
+ * @param portName port name
+ */
+function indexDataAndFireChange(port, portName, ports) {
+    return _.extend({
+        ondata: _.compose(
+            _.partial(fireChangeEvent, _.keys(ports)),
+            _.partial(indexPayload, createIndexBy(port), portName)
+        )
+    }, port);
+}
+
+/**
+ * Higher level function to extract the key from a payload given a path or property name.
+ * @param path JSON Pointer path or property name (that does not start with '/')
+ * @param object to extract the key from
+ */
+function createIndexBy(port) {
+    if (_.isFunction(port.indexBy))
+        return port.indexBy;
+    else if (_.isString(port.indexBy) && port.indexBy.charAt(0) == '/')
+        return jsonpointer.compile(port.indexBy).get;
+    else if (_.isString(port.indexBy))
+        return _.property(port.indexBy);
+}
+
+/**
+ * Higher level function to index payload by a given key.
+ * @this Port that the payload was received on
+ * @param indexBy property name, path, or function to index the payload by
+ * @param portName the port name of the payload to index
+ * @param payload to index
+ */
+function indexPayload(indexBy, portName, payload) {
+    // determine a key string or undefined if no key should be used
+    var key = indexBy == null ? undefined : ('' + indexBy.call(this, payload));
+    var rpf = this.nodeInstance.rpf = this.nodeInstance.rpf || {};
+    var payloads = rpf.payloads || {};
+    var index = key == null ? payload :
+        _.extend(payloads[portName] || {}, _.object([key], [payload]));
+    rpf.payloads = _.extend(payloads, _.object([portName], [index]));
+    var vni = rpf.vni('').inputState( portName, 'test', '*', rpf.payloads );
```

Comment:
Changes to this fill should be make in the change-component-factory PR, not this one.